### ISSUE TITLE
UART over USB / UART App

### ISSUE THREAD

**Issue by DSchndr** [Omega-Numworks/Omega#284]
#### What I want to see in the next version of Omega
It would be nice to be able to change between the internal uart and usb port uart in order to interface external electronics.

Also an app (& python functions) to change which UART gets used, change which baudrate it has and send/receive data would be nice to have.

This has been done before by zardam and zhaofengli, but needs to be updated (i did this some time ago, but it is also outdated)
https://github.com/zhaofengli/epsilon/tree/uart

**Comment by RedGl0w** [comment]
UART over usb is not anymore possible on n0110, but Can bus is

**Comment by DSchndr** [comment]
We could bit-bang uart on the n0110.
CAN would be nice but it requires an external transceiver (max3059 for example) in order to work.

**Comment by stevetuc** [comment]
> UART



> UART over usb is not anymore possible on n0110, but Can bus is

Why is that?

**Comment by RedGl0w** [comment]
> We could bit-bang uart on the n0110.
> CAN would be nice but it requires an external transceiver (max3059 for example) in order to work.
> 

With some passive components it could be possible if I remember well (but yes, we could also have some sort of software Serial arduino like)

**Comment by jabrena** [comment]
Any progress in this issue? Numworks could process sensor data from an Arduino board

**Comment by DSchndr** [comment]
If you have a N0100 you can use the uart on the usb port

Uart support gets somewhere disabled in the build system jungle so you would need to enable that and you would need to disable usb (there is some ion func) when you want to use uart
To this add patches from zardam and zhaofengli in order to have the app and support in python

Somewhere i had a recent version of omega with the uart app working, i dunno where the backup of it is tho

**Comment by jabrena** [comment]
I have a N0110 model to test it

**Comment by DSchndr** [comment]
**If you have a N0100...**
On N0110 you would have to write code for ion that does uart bitbang because the peripheral does not go to the usb data lines (but can be used as gpio iirc)

**Comment by RedGl0w** [comment]
> **If you have a N0100...**
> On N0110 you would have to write code for ion that does uart bitbang because the peripheral does not go to the usb data lines (but can be used as gpio iirc)

Yep, that was what I was talking about when I said "software serial like".
It would maybe not be super interesting (the calculator would drastically run slower, and you have other possibilities with these pins, like USB host / peripheral and can)

**Comment by jabrena** [comment]
Any progress on this?
